app/src/main/java/org/xbmc/kore/utils/Task.java

```diff
@@ -0,0 +1,81 @@
+package org.xbmc.kore.utils;
+
+import android.support.annotation.NonNull;
+
+/**
+ * Represents a computation that will eventually complete.
+ *
+ * @param <T> The type of the value being computed or awaited.
+ */
+public interface Task<T> {
+
+    /**
+     * Starts the computation, invokes a procedure when done.
+     *
+     * @param then The consumer of the value being computed.
+     */
+    void start(@NonNull OnFinish<? super T> then);
+
+    /**
+     * A procedure that awaits a value.
+     *
+     * @param <T> The type of the value generated by a previous computation.
+     */
+    interface OnFinish<T> {
+        void got(T result);
+    }
+
+    /**
+     * A sequence of tasks chained together.
+     *
+     * The values generated by the individual tasks are ignored except for the
+     * last one which may be accessed in the consumer if all of the tasks in
+     * the chain completed successfully.
+     *
+     * @param <T> The type of the value generated by the last task.
+     */
+    class Sequence<T> implements Task<T> {
+        private final Task<T> task;
+
+        /**
+         * Starts a sequence.
+         *
+         * @param task The last/outermost task
+         */
+        public Sequence(Task<T> task) {
+            this.task = task;
+        }
+
+        /**
+         * Continues the chain.
+         *
+         * @param next The next task to execute
+         * @param <U> The type of the value generated by the next task
+         * @return a sequence
+         */
+        public <U> Sequence<U> then(final Task<U> next) {
+            return new Sequence<>(new Task<U>() {
+                @Override
+                public void start(@NonNull final OnFinish<? super U> then) {
+                    task.start(new OnFinish<T>() {
+                        @Override
+                        public void got(T unused) {
```

Comment:
As I understand it, within a Monad the bind operator ensures the next operation is not performed when the previous failed. Therefore a programmer using a Monad does not need to worry about the logic and can simply assume if his operation fails the next in line won't be called. This is not the same as how it is implemented here. Here, the programmer has to know that he/she should call got(...) to make sure the next in line is executed or should NOT call got(...) to stop the chain. 
So we need something to indicate failure/success and I think the easiest way to do this, is by changing the `got(...)` method to `got(T result, boolean status)` and use the status to determine if `then.start(then)` should be called.